### PR TITLE
transition to using new bucket model version

### LINKED ISSUE #118: Update coupler AMIP files to work with newer version of Bucket model
The bucket model was improved to allow for snow on the surface, as well as to solve a multi layer heat equation. The design doc for the bucket model has been updated:https://www.overleaf.com/project/625c5f7677155b314c98eb75, and this model has been added to ClimaLSM.jl.

The next step is to update the coupler to use this version, rather than the previous version.

### PR DESCRIPTION
**Pull request by kmdeck**
# PULL REQUEST

## Purpose and Content
This PR updates the coupler driver file and other helper files so that the coupled runs work with the new version of the bucket model (with snow, and with a multi layer heat equation).

To do: followup on https://github.com/CliMA/ClimaCore.jl/issues/943
## Benefits and Risks
The benefit of this is that our AMIP model will be slightly more realistic.

## Linked Issues
(Provide references to any link issues. Use closes #issuenum to automatically close an open issue)
- Fixes #118 
- Closes #118 

## PR Checklist
- [X] This PR has a corresponding issue OR is linked to an SDI.
- [X] I have followed CliMA's codebase [contribution](https://clima.github.io/ClimateMachine.jl/latest/Contributing/) and [style](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) guidelines OR N/A.
- [X] I have followed CliMA's [documentation policy](https://github.com/CliMA/policies/wiki/Documentation-Policy).
- [X] I have checked all issues and PRs and I certify that this PR does not duplicate an open PR.
- [x] I linted my code on my local machine prior to submission OR N/A.
- [X] Unit tests are included OR N/A.
- [X] Code used in an integration test OR N/A.
- [x] All tests ran successfully on my local machine OR N/A.
- [X] All classes, modules, and function contain docstrings OR N/A.
- [X] Documentation has been added/updated OR N/A.

### CODE REVIEW

File: experiments/AMIP/moist_mpi_earth/bucket/bucket_init.jl [hunk 1]
```diff
@@ -157,18 +190,112 @@ function get_land_q(slab_sim::BucketSimulation, _...)
...
+            bucket_sim.params.ρc_soil .* mean(bucket_sim.integrator.u.bucket.T[colidx]) .* bucket_sim.domain.soil_depth
+    end
+
+    e_per_area .+=
+        -LSMP.LH_f0(bucket_sim.params.earth_param_set) .* LSMP.ρ_cloud_liq(bucket_sim.params.earth_param_set) .*
+        bucket_sim.integrator.u.bucket.σS
+    return e_per_area
+end
+
+"""
```
Comment (kmdeck) [inline experiments/AMIP/moist_mpi_earth/bucket/bucket_init.jl:215]: can remove as we dont support the land model being a slab anymore?

File: experiments/AMIP/moist_mpi_earth/cli_options.jl [hunk 2]
```diff
@@ -17,19 +17,19 @@ function parse_commandline()
         "--mode_name"
         help = "Mode of coupled simulation. [`amip`, `slabplanet`]"
```
Comment (kmdeck) [inline experiments/AMIP/moist_mpi_earth/cli_options.jl:18]: will revert when ready to merge

File: experiments/AMIP/moist_mpi_earth/coupler_driver.jl [hunk 3]
```diff
@@ -20,7 +20,7 @@ mode_name = parsed_args["mode_name"]
 energy_check = parsed_args["energy_check"]
 const FT = parsed_args["FLOAT_TYPE"] == "Float64" ? Float64 : Float32
 land_sim_name = "bucket"
-t_end = FT(time_to_seconds(parsed_args["t_end"]))
+t_end = FT(10000)#FT(time_to_seconds(parsed_args["t_end"]))
```
Comment (kmdeck) [inline experiments/AMIP/moist_mpi_earth/coupler_driver.jl:23]: will revert when ready to merge

File: experiments/AMIP/moist_mpi_earth/bucket/bucket_init.jl [hunk 4]
```diff
@@ -180,24 +151,29 @@ function bucket_init(::Type{FT}, tspan::Tuple{FT, FT}; space, dt::FT, saveat::FT
...
+    args = (params, CoupledAtmosphere{FT}(), CoupledRadiativeFluxes{FT}(), domain)
     model = BucketModel{FT, typeof.(args)...}(args...)
 
     # Initial conditions with no moisture
-    Y, p, coords = initialize(model, space)
+    Y, p, coords = initialize(model)
     anomaly = true
     hs_sfc = false
-    Y.bucket.T_sfc = map(coords) do coord
+    Y.bucket.T = map(coords.subsurface) do coord
```
Comment (kmdeck) [inline experiments/AMIP/moist_mpi_earth/bucket/bucket_init.jl:176]: @LenkaNovak right now the anomaly is 0 - do we want to change this?

**Comment by kmdeck** [comment]
FWIW, longer runs seem to run into this now:
-----------------------------------------
maxiter reached in saturation_adjustment:
    Method=NewtonsMethod, e_int=-5.53005e6, ρ=-0.0005846755, q_tot=0.0, T=-7435.205, maxiter=8, tol=71.7506
ERROR: TaskFailedException

File: experiments/AMIP/moist_mpi_earth/coupler_utils/flux_calculator.jl [hunk 5]
```diff
@@ -47,11 +47,12 @@ function calculate_surface_fluxes_atmos_grid!(integrator, info_sfc)
     FT = eltype(integrator.u.c.ρ)
     thermo_params = CAP.thermodynamics_params(integrator.p.params)
     surface_flux_params = CAP.surface_fluxes_params(integrator.p.params)
+    uₕ_int = Geometry.UVVector.(Spaces.level(Y.c.uₕ, 1))
```
Comment (LenkaNovak) [inline experiments/AMIP/moist_mpi_earth/coupler_utils/flux_calculator.jl:50]: Doesn't this allocate more than previously?

**Review by LenkaNovak — APPROVED** [review]
Looks good. Only a small comment left in the code. Thanks!

File: experiments/AMIP/moist_mpi_earth/coupler_utils/flux_calculator.jl (hunk 5 above)
Comment (kmdeck) [reply experiments/AMIP/moist_mpi_earth/coupler_utils/flux_calculator.jl]: Ill revert it back

**Pull request merged**